### ISSUE TITLE
Find a real solution to locking issue

### ISSUE THREAD

**Issue by felixdoerre** [felixdoerre/primus_vk#3]
Primus-VK can deadlock on CreateSwapchain.
The best way to circumvent this possible deadlock is probably to remove the dependency on calling vk-API functions directly.

**Comment by alois31** [comment]
I think we should try to get rid of the need to acquire the global lock before we join `cod` (lines 666-676 in primus_vk.cpp, in PrimusVK_CreateSwapchainKHR)
```
{
    scoped_lock l(global_lock); // <-----
    if(cod == nullptr){
      std::cerr << "no thread to join\n";
      return VK_ERROR_INITIALIZATION_FAILED;
    }
    TRACE("joining secondary device creation");
    TRACE("When startup hangs here, you have probably hit the initialization deadlock");
    cod->join();
    TRACE("joining succeeded. Luckily initialization deadlock did not occur.");
}
```
Actually, I don't know whether this is needed at all, but probably libvulkan does some weird callbacks I'm not aware of.

**Comment by felixdoerre** [comment]
I am not sure if this is a good idea. I'll probably have to re-think the whole locking inside the layer. I kept layer-wide locking as this is also present in the sample layer. I'll take a closer look, when the issue in libvulkan is resolved.

The root cause for this problem is in libvulkan, that does mutex-locking of the whole vulkan api: e.g. 
https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/trampoline.c#L762
I am working with a Vulkan-Loader developer to resolve this issue. He told me that the correct way to go about this is to not link to the loader directly (which is exactly what I tried to suggest in the issue description).
I am discussing this issue (and a solution) in https://github.com/KhronosGroup/Vulkan-Loader/issues/74.
The whole discussion looks promising and I have submitted a pull request that should resolve this issue, however it will probably still take a few weeks until I get more review and this may be merged into vulkan (as it slightly changes the public interface of the loader).

**Comment by alois31** [comment]
Sorry, I wasn't quite clear in my comment.

The deadlock, at least for me, does *not* happen in libvulkan, but in our layer. This happens because we are acquiring our global lock in PrimusVK_CreateSwapchainKHR, which, when the main thread is "too fast", happens before the secondary device is set up (what we are waiting for in PrimusVK_CreateSwapchainKHR). Here, we are deadlocked, as the secondary thread can never acquire our global lock again.

Note that we are *not* calling any vk-API functions while we are holding the lock in PrimusVK_CreateSwapchainKHR: We only output the tracing messages and join a thread.

But I agree that we should not call vk-API functions anyway, as that may cause crashes (as outline in your referenced issue for Vulkan-Loader, which was a very insightful read).

**Comment by felixdoerre** [comment]
Aah, interesting. So I obviously didn't really try that out. I did now and apparently removing that lock also does solve the problem on my system. I (wrongly) assumed vkCreateSwapchainKHR (https://github.com/KhronosGroup/Vulkan-Loader/blob/0366721b5bbbc61490ee5a2cb5120b6c6d06e98c/loader/wsi.c#L385) also fetches the Vulkan loader lock as `vkCreateDevice` (https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/trampoline.c#L762). However it does not. So yes, the deadlock can be solved there. But pursuing the path that could remove linking to the loader is still good.

I'll remove that lock in `vkCreateSwapchainKHR`.

**Comment by felixdoerre** [comment]
This issue should be solved by the commit series around 134535c4d60cc1d3271d825ade08470540af539b.
